src/ledger/LedgerTxn.cpp

```diff
@@ -2566,6 +2620,39 @@ LedgerTxn::Impl::EntryIteratorImpl::clone() const
     return std::make_unique<EntryIteratorImpl>(mIter, mEnd);
 }
 
+// Implementation of ThreadInvariant --------------------------
+ThreadInvariant::ThreadInvariant()
+    : mThreadID(std::this_thread::get_id()), mActiveThreadIdSet(false)
+{
+}
+
+void
+ThreadInvariant::throwIfWrongThread() const
+{
+    std::lock_guard<std::recursive_mutex> guard(mThreadInvariantMutex);
+    if (mActiveThreadIdSet && mThreadID != std::this_thread::get_id())
+    {
+        throw std::runtime_error("ThreadInvariant called from wrong thread");
+    }
+}
```

Comment:
I'm not complaining about this, but curious: taking a mutex and doing a TLS access on every single method in this class _might_ show up in profiles, considering how hot the ltx sometimes is; does it?